File: loop_rooms.py
from sys import argv

#open and parse file
#M,N are 2 integers: number of rows and columns
#maze is a list of strings
with open(argv[1]) as f:
    M,N=map(int,(f.readline()).split())
    maze=[x.strip() for x in f.readlines()]

#find left and right immediate exits, mark them with '1'
#find up and down immediate exits, mark them with '1'
#maze is now a 2d list
maze=[['1' if ((c=='L'and j==0) or (c=='R' and j==N-1)) else c for j,c in enumerate(row)] for row in maze]
maze=[['1' if c=='U' else c for c in row] if i==0 else ['1' if c=='D' else c for c in row] if i==M-1 else row for i,row in enumerate(maze)]

#function that checks the path starting from grid[r][c]
#follows the path and stores each passing tile in set current
#all tiles in current set to '2'
#if an exit '1' is reached set all tiles in path to '1'
#if a loop '0' is reached or loop on self '2' then set all tiles in path to '0'
def path(grid,r,c):
    #set of tiles in path starting from grid[r][c]
    current=set()
    while 1:
        #add tile to path
        current.add((r,c))

        if grid[r][c]=='L':
            if grid[r][c-1]=='1':
                for a in current:
                    grid[a[0]][a[1]]='1'
                return
            elif grid[r][c-1]=='2' or grid[r][c-1]=='0':
                for a in current:
                    grid[a[0]][a[1]]='0'
                return
            else:
                grid[r][c]='2'
                c=c-1

        elif grid[r][c]=='R':
            if grid[r][c+1]=='1':
                for a in current:
                    grid[a[0]][a[1]]='1'
                return
            elif grid[r][c+1]=='2' or grid[r][c+1]=='0':
                for a in current:
                    grid[a[0]][a[1]]='0'
                return
            else:
                grid[r][c]='2'
                c=c+1

        elif grid[r][c]=='U':
            if grid[r-1][c]=='1':
                for a in current:
                    grid[a[0]][a[1]]='1'
                return
            elif grid[r-1][c]=='2' or grid[r-1][c]=='0':
                for a in current:
                    grid[a[0]][a[1]]='0'
                return
            else:
                grid[r][c]='2'
                r=r-1

        elif grid[r][c]=='D':
            if grid[r+1][c]=='1':
                for a in current:
                    grid[a[0]][a[1]]='1'
                return
            elif grid[r+1][c]=='2' or grid[r+1][c]=='0':
                for a in current:
                    grid[a[0]][a[1]]='0'
                return
            else:
                grid[r][c]='2'
                r=r+1

        #return immediately if grid[r][c] is already '0' or '1'
        else:
            return

#call path for every tile
for i in range(M):
    for j in range(N):
        path(maze,i,j)

#count all '0'
looprooms=sum(row.count('0') for row in maze)

print(looprooms)

File: loop_rooms.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
using namespace std;

//recursively checks if neighbouring tiles connect to input tile and updates t
//Parametres:
/*t: bool grid
m,n: coordinates of tile that called
D: maze
M,N: dimensions of maze*/
void neighbours(bool** t, int m, int n, string D[], int M, int N) {
    if(m!=0) {
        if(D[m-1][n]=='D' && t[m-1][n]!=1) {
            t[m-1][n]=1;
            neighbours(t,m-1,n,D,M,N);
        }
    }
    if(m!=M-1) {
        if(D[m+1][n]=='U' && t[m+1][n]!=1) {
            t[m+1][n]=1;
            neighbours(t,m+1,n,D,M,N);
        }
    }
    if(n!=0) {
        if(D[m][n-1]=='R' && t[m][n-1]!=1) {
            t[m][n-1]=1;
            neighbours(t,m,n-1,D,M,N);
        }
    }
    if(n!=N-1) {
        if(D[m][n+1]=='L' && t[m][n+1]!=1) {
            t[m][n+1]=1;
            neighbours(t,m,n+1,D,M,N);
        }
    }
    return;
}

int main(int argc, char** argv) {
    //open file
    ifstream infile;
    infile.open(argv[1]);
    if(!infile) {
        cout << "Error opening " << argv[1] << "\n";
        exit(-1);
    }

    //get data from file
    int M=0, N=0;
    string str;
    stringstream ss(str);

    getline(infile, str, ' ');
    ss<<str;
    ss>>M;
    ss.str("");
    ss.clear();

    getline(infile,str);
    ss<<str;
    ss>>N;
    ss.str("");
    ss.clear();

    string D[M];
    for(int i=0; i<M; ++i) {
        getline(infile,D[i]);
    }

    //print maze
    /*cout << M << " " << N << "\n";
    for(int i=0; i<M; ++i) {
        cout << D[i] << '\n';
    }
    cout << '\n';*/

    //creating bool grid to track which tiles you can exit from
    bool **t;
    t = new bool *[M];
    for(int i = 0; i <M; i++)
        t[i] = new bool[N];

    //find upper and lower immediate exits
    //set them to 1
    for(int i=0; i<N; i++) {
        if(D[0][i]=='U') {
            t[0][i]=1;
        }
        if(D[M-1][i]=='D') {
            t[M-1][i]=1;
        }
    }
    //find left and right immediate exits
    //set them to 1
    for(int i=0; i<M; i++) {
        if(D[i][0]=='L') {
            t[i][0]=1;
        }
        if(D[i][N-1]=='R') {
            t[i][N-1]=1;
        }
    }

    //from each immediate exit call function neighbours
    //if you can exit from a tile, if its neighbours lead to it you can exit
    //from them too
    for(int i=0; i<N; i++) {
        if(t[0][i]==1) {
            neighbours(t,0,i,D,M,N);
        }
        if(t[M-1][i]==1) {
            neighbours(t,M-1,i,D,M,N);
        }
    }
    for(int i=0; i<M; i++) {
        if(t[i][0]==1) {
            neighbours(t,i,0,D,M,N);
        }
        if(t[i][N-1]==1) {
            neighbours(t,i,N-1,D,M,N);
        }
    }

    //all tiles you can't exit from have 0, so count them in bad
    int bad=0;
    for(int i=0; i<M; ++i) {
        for(int j=0; j<N; ++j) {
            if(t[i][j]==0) {
                bad++;
            }
        }
    }

    cout << bad << '\n';
}
